libs/qa/dataset_summary.py

```diff
@@ -0,0 +1,139 @@
+from typing import Optional, Tuple
+
+import pydantic
+import pandas as pd
+import numpy as np
+from covidactnow.datapublic.common_fields import CommonFields
+from libs.datasets.dataset_utils import AggregationLevel
+from libs.datasets.timeseries import TimeseriesDataset
+from libs.datasets import combined_datasets
+
+IGNORE_COLUMNS = [
+    CommonFields.STATE,
+    CommonFields.COUNTRY,
+    CommonFields.COUNTY,
+    CommonFields.AGGREGATE_LEVEL,
+]
+
+VARIABLE_FIELD = "variable"
+
+
+class TimeseriesSummary(pydantic.BaseModel):
+    """Summary of timeseries dataset at a given commit sha."""
+
+    sha: str
+    timeseries: TimeseriesDataset
+    summary: pd.DataFrame
+    fips: Optional[str]
+    level: Optional[AggregationLevel]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+def generate_field_summary(series: pd.Series) -> pd.Series:
+
+    has_value = not series.isnull().all()
+    min_date = None
+    max_date = None
+    max_value = None
+    min_value = None
+    latest_value = None
+    num_observations = 0
+    largest_delta = None
+    largest_delta_date = None
+
+    if has_value:
+        min_date = series.first_valid_index()[1]
+        max_date = series.last_valid_index()[1]
+        latest_value = series[series.notnull()].iloc[-1]
+        max_value = series.max()
+        min_value = series.min()
+        num_observations = len(series[series.notnull()])
+        largest_delta = series.diff().abs().max()
+        # If a
+        if len(series.diff().abs().dropna()):
+            largest_delta_date = series.diff().abs().idxmax()[1]
+
+    results = {
+        "has_value": has_value,
+        "min_date": min_date,
+        "max_date": max_date,
+        "max_value": max_value,
+        "min_value": min_value,
+        "latest_value": latest_value,
+        "num_observations": num_observations,
+        "largest_delta": largest_delta,
+        "largest_delta_date": largest_delta_date,
+    }
+    return pd.Series(results)
+
+
+def summarize_timeseries_fields(data: pd.DataFrame) -> pd.DataFrame:
+    data = data[[column for column in data.columns if column not in IGNORE_COLUMNS]]
+
+    melted = pd.melt(data, id_vars=[CommonFields.FIPS, CommonFields.DATE]).set_index(
+        [CommonFields.FIPS, CommonFields.DATE, VARIABLE_FIELD]
+    )
+    fips_variable_grouped = melted.groupby([CommonFields.FIPS, VARIABLE_FIELD])
+    return fips_variable_grouped["value"].apply(generate_field_summary).unstack()
+
+
+def find_missing_values_in_summary(
+    summary_l: pd.DataFrame, summary_r: pd.DataFrame
+) -> pd.DataFrame:
+    """Returns values that are in the left summary but not in right summary."""
```

Comment:
this looks like it depends on the two dataframes having the same index, or can you be sure there is a summary_r.has_value False for every input that doesn't have a value?